File: crates/aegis-http/src/cookies.rs
use axum::http::header::SET_COOKIE;
use axum::http::{HeaderMap, HeaderValue};
use aegis_config::{CookieConfig, SameSite};
use time::Duration;

pub fn set_session_cookie(
    headers: &mut HeaderMap,
    cookie: &CookieConfig,
    token: &str,
    max_age: Duration,
) {
    let age_secs = max_age.whole_seconds().max(0);
    let mut parts = vec![
        format!("{}={token}", cookie.name),
        format!("Path={}", cookie.path),
        format!("Max-Age={age_secs}"),
        format!(
            "SameSite={}",
            match cookie.same_site {
                SameSite::Strict => "Strict",
                SameSite::Lax => "Lax",
                SameSite::None => "None",
            }
        ),
    ];

    if cookie.http_only {
        parts.push("HttpOnly".to_owned());
    }
    if cookie.secure {
        parts.push("Secure".to_owned());
    }
    if let Some(domain) = &cookie.domain {
        parts.push(format!("Domain={domain}"));
    }

    let value = parts.join("; ");
    if let Ok(hv) = HeaderValue::from_str(&value) {
        headers.append(SET_COOKIE, hv);
    }
}

pub fn clear_session_cookie(headers: &mut HeaderMap, cookie: &CookieConfig) {
    let mut parts = vec![
        format!("{}=", cookie.name),
        format!("Path={}", cookie.path),
        "Max-Age=0".to_owned(),
        format!(
            "SameSite={}",
            match cookie.same_site {
                SameSite::Strict => "Strict",
                SameSite::Lax => "Lax",
                SameSite::None => "None",
            }
        ),
    ];

    if cookie.http_only {
        parts.push("HttpOnly".to_owned());
    }
    if cookie.secure {
        parts.push("Secure".to_owned());
    }
    if let Some(domain) = &cookie.domain {
        parts.push(format!("Domain={domain}"));
    }

    let value = parts.join("; ");
    if let Ok(hv) = HeaderValue::from_str(&value) {
        headers.append(SET_COOKIE, hv);
    }
}

File: sdk/packages/aegis-server/src/next/cookies.ts
import { cookies } from "next/headers";
import type { NextResponse } from "next/server";

export const DEFAULT_AEGIS_SESSION_COOKIE_NAME = "aegis_session";

export type AegisCookieStore = {
  get(name: string): { value: string } | undefined;
};

export type AegisSessionCookieOptions = {
  name?: string;
  path?: string;
  domain?: string;
  secure?: boolean;
  httpOnly?: boolean;
  sameSite?: "lax" | "strict" | "none";
  maxAge?: number;
};

export async function getCookieStore(): Promise<AegisCookieStore> {
  return cookies();
}

export function getSessionCookie(
  store: AegisCookieStore,
  name: string = DEFAULT_AEGIS_SESSION_COOKIE_NAME,
): string | undefined {
  return store.get(name)?.value;
}

export function setSessionCookie(
  response: NextResponse,
  value: string,
  options: AegisSessionCookieOptions = {},
): NextResponse {
  response.cookies.set({
    name: options.name ?? DEFAULT_AEGIS_SESSION_COOKIE_NAME,
    value,
    path: options.path ?? "/",
    ...(options.domain !== undefined ? { domain: options.domain } : {}),
    secure: options.secure ?? true,
    httpOnly: options.httpOnly ?? true,
    sameSite: options.sameSite ?? "lax",
    ...(options.maxAge !== undefined ? { maxAge: options.maxAge } : {}),
  });

  return response;
}

export function clearSessionCookie(
  response: NextResponse,
  options: AegisSessionCookieOptions = {},
): NextResponse {
  response.cookies.set({
    name: options.name ?? DEFAULT_AEGIS_SESSION_COOKIE_NAME,
    value: "",
    path: options.path ?? "/",
    ...(options.domain !== undefined ? { domain: options.domain } : {}),
    secure: options.secure ?? true,
    httpOnly: options.httpOnly ?? true,
    sameSite: options.sameSite ?? "lax",
    maxAge: 0,
  });

  return response;
}
